### PR TITLE
Remove NULL return in GetFE

### PR DESCRIPTION
**Pull request by kmittal2**
| PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3613](https://github.com/mfem/mfem/pull/3613) | @kmittal2 | @tzanio | @v-dobrev + @dylan-copeland | 8/12/23 | 8/27/23 | 9/3/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by tzanio** [comment]
Are we sure we don't rely on the NULL return value somewhere in MFEM or one of its applications?

Maybe it is OK to force an error in that case, but I'd like to make sure we understand the implications of this change.

Thoughts @mfem/developers ?

File: fem/fespace.cpp [hunk 1]
```diff
@@ -2786,9 +2786,10 @@ FiniteElementSpace::GetElementDofs(int elem, Array<int> &dofs) const
 
 const FiniteElement *FiniteElementSpace::GetFE(int i) const
 {
-   if (i < 0 || !mesh->GetNE()) { return NULL; }
-   MFEM_VERIFY(i < mesh->GetNE(),
-               "Invalid element id " << i << ", maximum allowed " << mesh->GetNE()-1);
+   MFEM_ASSERT(mesh->GetNE(),"Empty Partition");
+   MFEM_VERIFY(i >= 0 && i < mesh->GetNE(),
```
Comment (dylan-copeland) [inline fem/fespace.cpp:2790]: This condition implies `mesh->GetNE() > 0`, so the above assertion is redundant.

Comment (kmittal2) [reply fem/fespace.cpp]: I refactored it to give more informative error.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/fespace.cpp [hunk 2]
```diff
@@ -2803,9 +2803,17 @@ FiniteElementSpace::GetElementDofs(int elem, Array<int> &dofs) const
...
 {
-   if (i < 0 || !mesh->GetNE()) { return NULL; }
-   MFEM_VERIFY(i < mesh->GetNE(),
-               "Invalid element id " << i << ", maximum allowed " << mesh->GetNE()-1);
+   if (i < 0 || i >= mesh->GetNE())
+   {
+      if (mesh->GetNE() == 0)
+      {
+         MFEM_ABORT("Empty MPI partition!!! NOTE: FESpace::GetFE has been "
+                    "updated to abort instead of returning NULL. The user must "
```
Comment (dylan-copeland) [inline fem/fespace.cpp:2811]: Just a minor suggestion, but I suppose after this is merged, the comment about the being updated may be confusing, if the reader does not know the history of this function. It seems simpler and more straightforward to say only that empty partitions are not permitted within this function. That makes the comment about the current state of the code, not the history.

**Comment by dylan-copeland** [comment]
@kmittal2 this PR makes a design change from one reasonable choice to another. I can imagine pros and cons for both choices, but I'm not sure what your motivation is for this change. Could you please provide some explanation of the motivation and impact of this change?

**Comment by kmittal2** [comment]
> @kmittal2 this PR makes a design change from one reasonable choice to another. I can imagine pros and cons for both choices, but I'm not sure what your motivation is for this change. Could you please provide some explanation of the motivation and impact of this change?

@dylan-copeland I am not sure if I see returning NULL as a reasonable choice since it can lead to segfaults which are not straightforward to track. 
We don't usually run empty partitons, but with SubMesh we don't have a way to avoid empty partitions as we don't currently support rebalancing of conforming meshes. As a result, folks using SubMesh are likely to run into this issue. For example, a quick search for `GetFE(0)->Get...` in the source shows several instances that would produce a segfault as a result of the current design.

**Comment by v-dobrev** [comment]
My understanding is that returning `NULL` has caused issues in practice which were not straightforward to track, so this should help with such issues. If some codes are affected, it will be an easy update, I think. 🤞

**Review by v-dobrev — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**